File: src/model/reference/rayleigh.zig
const std = @import("std");

const fraction_n2 = 78.084;
const fraction_o2 = 20.946;
const fraction_ar = 0.934;
const fraction_co2 = 0.036;
const reference_number_density_cm3 = 2.5468993e19;

fn kingFactorN2(wavelength_nm: f64) f64 {
    const sigma_um_inv = 1000.0 / @max(wavelength_nm, 1.0);
    return 1.034 + 3.17e-4 * sigma_um_inv * sigma_um_inv;
}

fn kingFactorO2(wavelength_nm: f64) f64 {
    const sigma_um_inv = 1000.0 / @max(wavelength_nm, 1.0);
    const sigma_sq = sigma_um_inv * sigma_um_inv;
    return 1.096 + 1.385e-3 * sigma_sq + 1.448e-4 * sigma_sq * sigma_sq;
}

fn kingFactorAir(wavelength_nm: f64) f64 {
    const weighted_sum =
        fraction_n2 * kingFactorN2(wavelength_nm) +
        fraction_o2 * kingFactorO2(wavelength_nm) +
        fraction_ar * 1.0 +
        fraction_co2 * 1.15;
    return weighted_sum / (fraction_n2 + fraction_o2 + fraction_ar + fraction_co2);
}

pub fn refractiveIndexDryAir(wavelength_nm: f64) f64 {
    const sigma_um_inv = 1000.0 / @max(wavelength_nm, 1.0);
    const sigma_sq = sigma_um_inv * sigma_um_inv;
    const refractivity =
        8060.51 +
        2480990.0 / (132.274 - sigma_sq) +
        17455.7 / (39.32957 - sigma_sq);
    return 1.0 + refractivity * 1.0e-8;
}

pub fn depolarizationFactorAir(wavelength_nm: f64) f64 {
    const king_factor_air = kingFactorAir(wavelength_nm);
    return 6.0 * (king_factor_air - 1.0) / (3.0 + 7.0 * king_factor_air);
}

pub fn crossSectionCm2(wavelength_nm: f64) f64 {
    const safe_wavelength_nm = @max(wavelength_nm, 1.0);
    const refractive_index = refractiveIndexDryAir(safe_wavelength_nm);
    const numerator = 24.0 * std.math.pi * std.math.pi * std.math.pi;
    const wavelength_cm = safe_wavelength_nm * 1.0e-7;
    var cross_section =
        numerator /
        std.math.pow(f64, wavelength_cm, 4.0) /
        (reference_number_density_cm3 * reference_number_density_cm3);
    cross_section *=
        std.math.pow(f64, refractive_index * refractive_index - 1.0, 2.0) /
        std.math.pow(f64, refractive_index * refractive_index + 2.0, 2.0);
    return cross_section * kingFactorAir(safe_wavelength_nm);
}

pub fn scatteringOpticalDepthForColumn(
    wavelength_nm: f64,
    air_column_density_cm2: f64,
) f64 {
    return crossSectionCm2(wavelength_nm) * @max(air_column_density_cm2, 0.0);
}

test "rayleigh cross section and depolarization match dry-air O2A-scale expectations" {
    try std.testing.expectApproxEqAbs(@as(f64, 0.02771347468938541), depolarizationFactorAir(760.0), 1.0e-12);
    try std.testing.expectApproxEqAbs(@as(f64, 1.213452195882859e-27), crossSectionCm2(760.0), 1.0e-39);
    try std.testing.expect(crossSectionCm2(405.0) > crossSectionCm2(760.0));
    try std.testing.expect(depolarizationFactorAir(405.0) > depolarizationFactorAir(760.0));
}

File: rust_src/input/reference/rayleigh.rs
const FRACTION_N2: f64 = 78.084;
const FRACTION_O2: f64 = 20.946;
const FRACTION_AR: f64 = 0.934;
const FRACTION_CO2: f64 = 0.036;
const REFERENCE_NUMBER_DENSITY_CM3: f64 = 2.5468993e19;

fn king_factor_n2(wavelength_nm: f64) -> f64 {
    let sigma_um_inv = 1000.0 / wavelength_nm.max(1.0);
    1.034 + 3.17e-4 * sigma_um_inv * sigma_um_inv
}

fn king_factor_o2(wavelength_nm: f64) -> f64 {
    let sigma_um_inv = 1000.0 / wavelength_nm.max(1.0);
    let sigma_sq = sigma_um_inv * sigma_um_inv;
    1.096 + 1.385e-3 * sigma_sq + 1.448e-4 * sigma_sq * sigma_sq
}

fn king_factor_air(wavelength_nm: f64) -> f64 {
    let weighted_sum = FRACTION_N2 * king_factor_n2(wavelength_nm)
        + FRACTION_O2 * king_factor_o2(wavelength_nm)
        + FRACTION_AR
        + FRACTION_CO2 * 1.15;
    weighted_sum / (FRACTION_N2 + FRACTION_O2 + FRACTION_AR + FRACTION_CO2)
}

pub fn refractive_index_dry_air(wavelength_nm: f64) -> f64 {
    let sigma_um_inv = 1000.0 / wavelength_nm.max(1.0);
    let sigma_sq = sigma_um_inv * sigma_um_inv;
    let refractivity = 8060.51 + 2480990.0 / (132.274 - sigma_sq) + 17455.7 / (39.32957 - sigma_sq);
    1.0 + refractivity * 1.0e-8
}

pub fn depolarization_factor_air(wavelength_nm: f64) -> f64 {
    let king_factor_air = king_factor_air(wavelength_nm);
    6.0 * (king_factor_air - 1.0) / (3.0 + 7.0 * king_factor_air)
}

pub fn cross_section_cm2(wavelength_nm: f64) -> f64 {
    let safe_wavelength_nm = wavelength_nm.max(1.0);
    let refractive_index = refractive_index_dry_air(safe_wavelength_nm);
    let numerator = 24.0 * std::f64::consts::PI.powi(3);
    let wavelength_cm = safe_wavelength_nm * 1.0e-7;
    let mut cross_section = numerator
        / wavelength_cm.powi(4)
        / (REFERENCE_NUMBER_DENSITY_CM3 * REFERENCE_NUMBER_DENSITY_CM3);
    cross_section *= (refractive_index * refractive_index - 1.0).powi(2)
        / (refractive_index * refractive_index + 2.0).powi(2);
    cross_section * king_factor_air(safe_wavelength_nm)
}

pub fn scattering_optical_depth_for_column(wavelength_nm: f64, air_column_density_cm2: f64) -> f64 {
    cross_section_cm2(wavelength_nm) * air_column_density_cm2.max(0.0)
}
